src/memory/AmdFlash.cc

```diff
@@ -35,152 +41,125 @@ AmdFlash::AmdFlash(const std::string& name, const ValidatedChip& validatedChip,
-			assert(rom == nullptr); // 'rom' remains nullptr
-			// only if an actual sha1sum was given, tell the user we failed to use it
-			if (initialContentSpecified) {
-				config.getCliComm().printWarning(
-					"Could not load specified initial content "
-					"for flash ROM: ", e.getMessage());
-			}
+			file = File(path, File::OpenMode::LOAD_PERSISTENT);
+		} catch (FileException& e) {
+			config.getCliComm().printWarning("Couldn't load Flash data ",
+				filename, " (", e.getMessage(), ").");
 		}
+	} catch (FileException& e) {
+		// SRAM file not found, assuming blank SRAM content.
 	}
 
-	readAddress.resize(numSectors);
-	auto romSize = rom ? rom->size() : 0;
-	size_t offset = 0;
-	for (size_t sector = 0; const AmdFlash::Region& region : chip.geometry.regions) {
-		for (size_t regionSector = 0; regionSector < region.count; regionSector++, sector++) {
-			auto sectorSize = region.size;
-			if (isSectorWritable(sector)) {
-				readAddress[sector] = &(*ram)[writeAddress[sector]];
-				if (!loaded) {
-					auto* ramPtr = const_cast<uint8_t*>(
-						&(*ram)[writeAddress[sector]]);
-					if (offset >= romSize) {
-						// completely past end of rom
-						ranges::fill(std::span{ramPtr, sectorSize}, 0xFF);
-					} else if (offset + sectorSize >= romSize) {
-						// partial overlap
-						auto last = romSize - offset;
-						auto missing = sectorSize - last;
-						const uint8_t* romPtr = &(*rom)[offset];
-						ranges::copy(std::span{romPtr, last}, ramPtr);
-						ranges::fill(std::span{&ramPtr[last], missing}, 0xFF);
-					} else {
-						// completely before end of rom
-						const uint8_t* romPtr = &(*rom)[offset];
-						ranges::copy(std::span{romPtr, sectorSize}, ramPtr);
-					}
-				}
+	size_t fileRemaining = file.is_open() ? file.getSize() : 0;
+	const std::span<const uint8_t> romData = rom ? std::span{*rom} : std::span<const uint8_t>{};
+	const std::span<uint8_t> ramData = ram.getWriteBackdoor();
+	for (size_t sector = 0, address = 0; const AmdFlash::Region& region : chip.geometry.regions) {
+		for (size_t regionSector = 0; regionSector < region.count; regionSector++, sector++, address += region.size) {
+			if (sectors[sector].modified) {
+				const size_t size = std::min<size_t>(region.size, fileRemaining);
+				if (size > 0) file.read(ramData.subspan(address, size));
+				ranges::fill(ramData.subspan(address + size, region.size - size), 0xFF);
+				fileRemaining -= size;
 			} else {
-				assert(rom); // must have rom constructor parameter
-				if ((offset + sectorSize) <= romSize) {
-					readAddress[sector] = &(*rom)[offset];
-				} else {
-					readAddress[sector] = nullptr;
-				}
+				const size_t size = std::min(address + region.size, romData.size()) - std::min(address, romData.size());
+				ranges::copy(romData.subspan(address, size), ramData.subspan(address, size));
+				ranges::fill(ramData.subspan(address + size, region.size - size), 0xFF);
 			}
-			offset += sectorSize;
 		}
 	}
-	assert(offset == size());
```

Comment:
If I understand correctly:
* The data-file only contains those sectors that have been modified.
* The meta-file indicates which sectors exactly that are.
* And if this meta-file is missing we assume all writable sectors have been modified. This is the scenario when upgrading from an older openMSX release.

Right?

--

In general we don't try to be super robust when loading savestates which have been manually edited by the users (so savestates which contain invalid data).

However in this case it's possible the user specified the wrong `-romtype` command line parameter, and then we may have a mismatch between the current `chip.geometry` and `std::vector sectors`. E.g. `sectors` may contain more or less number of elements than the number of sectors in `geometry`.

Maybe we should add a check (around line 69, where we do `in.serialize(...)`) that the loaded number of sectors matches the expected amount. (I'm not sure what to do in case of a mismatch).

Also if the number of sectors does match, the geometry may still be different. I think the loading code itself is robust (e.g. it won't read beyond the end of the file). But the data may still end up in the wrong places.

Maybe we should extend the meta info with a field indicating the chip ID?

Reply:
> However in this case it's possible the user specified the wrong `-romtype` command line parameter, and then we may have a mismatch between the current `chip.geometry` and `std::vector sectors`. E.g. `sectors` may contain more or less number of elements than the number of sectors in `geometry`.

Hmm very good point, I hadn’t thought of that.

Perhaps rather than store the modified sectors, it’d be better to store the ranges of memory that were modified (rounded to whole sectors), and when loading it would just mark the relevant sectors as modified. That wouldn’t be too difficult to implement, and it’d be more robust since it relates the the physical location in memory which is related to the ROM rather than the mapper.

That might be nicer than adding chip ID to the metadata, because then I don’t know what a good fallback behaviour would be (as you remark as well). Whereas with this even if the FlashROM geometry is different it will just remap the modified regions to that and mark the appropriate sectors modified.

Reply:
A good goal might be:
* If a flash-rom has already been written to (e.g. a savegame), and the user accidentally uses this data in combination with a different (wrong?) rom-type, then this shouldn't result in removing/corrupting the existing savegame.

I think your suggestion (using absolute offsets rather than sector numbers) already comes closer. But it might still loose data. For example in the scenario where the 2nd (wrong) mapper-type is a smaller flash rom, and the modified data is located past the size of the smaller mapper.

A second scenario that we may want to consider is:
* You initially start the game with the wrong mapper-type.
* Later you correct this mistake (= switch to the correct mapper).
* It should be possible to recover from this. So end up in the state that is identical to as-if you immediately used to correct type.

I'm not sure how to fully achieve this second goal.

Maybe we could:
* Do store the flashrom-id in the meta-data.
* Compare this ID when we use the data.
* If there's a mismatch, make the full flash read-only And print a warning with the full path of the file that should be deleted to recover (if the new mapper-type is actually the correct one).

When the data is stored in a single file, it's easier to tell the user what file to delete.

Reply:
Also if you store ranges of modified data (rounded to sector boundaries). But if you then load this data in a different flash-rom. These old boundaries may not map to new boundaries. And what will we do then? Extend again to sector boundaries? But that will result in problems when we correct our mistake and switch back to the original mapper (now we unintentionally extended the number of written-to sectors).

Reply:
Another option;

I could encode the manufacturer / device ID in the filename, like `m01-d1000.flash`. Then if the device is different for some reason you will get a separate persistent file. But if the user is certain they are compatible and wants to carry over the data, they can simply rename the file.

Reply:
I notice by the way that the manufacturer / device ID combination is not a unique identifier, and does not guarantee that they support the same command set;

The AM29F040 and AM29F040B both are identified as 01 / A4, but the latter supports status DQ2 and programming in erase suspend mode which the first does not.

The AM29F016, AM29F016B and AM29F016D are all identified as 01 / AD, but the D version supports CFI and Unlock Bypass which the other two do not.

I do think it can be assumed that devices with the same ID have the same geometry though. These are all iterations on the same design, coinciding with process node upgrades from 0.5 µm to 0.32 µm (B) to 0.23 µm (D).

Reply:
Yes, in this context (different filenames for different mappers) the geometry should be sufficient, I think.